3주차 과제. TEST

EunjinJwa opened the pull request:
과제 3(진행중) PR 올립니다. 

강의를 보면서 WebMvcTest, AutoConfigureMockMvc 각 Mvc 테스트가 어떻게 다른지 확인하기 위해 각각 테스트코드 작성해서 다른점을 확인해보았습니다. 

TaskControllerWebMvcTest
- @WebMvcTest
- TaskService 처리를 @MockBean으로 생성하여 조회 위주로 테스트 코드 작성

TaskControllerMockTest
- @SpringBootTest, @AutoConfigureMockMvc
- create, update 및 전체적인 CRUD 시나리오 테스트 작성

그럼 잘부탁드립니다 :)

EunjinJwa committed:
TaskController UnitTest 작성

EunjinJwa committed:
controller web test 추가 - 작성중

EunjinJwa committed:
mockMvc TEST 코드 작성

EunjinJwa committed:
불필요한 코드 제거

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMockTest.java:
```diff
@@ -0,0 +1,103 @@
+	public void list() throws Exception {
+		mockMvc.perform(get("/tasks"))
+				.andExpect(status().isOk())
+				.andDo(print());
```
응답결과를 출력해서 확인하셨던 것 같습니다. 눈으로 결과를 확인하고 어떻게 테스트 할지 좀 살펴보셨을 것 같아요. 눈으로 확인하지 않고 확인할 수 있도록 검증도 작성해보면 좋을 것 같습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerUnitTest.java:
```diff
@@ -0,0 +1,115 @@
+
+	@Test
+	public void createTask() {
+		int before = taskController.list().size();
+		taskController.create(new Task());
+		int after = taskController.list().size();
+
+		assertThat(after - before).isEqualTo(1);
```
테스트 대상을 실행하면 어떤 변화가 일어나는지 테스트를 작성해 주셨네요. 앞으로도 테스트를 작성할 때, 어떤 변화가 일어나는지에 집중하여 테스트를 작성하면 왜 이것을 사용해야 하는지 의도를 드러내기 쉬울거예요

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMockTest.java:
```diff
@@ -0,0 +1,103 @@
+	}
+
+	@Test
+	public void scenarioTest() throws Exception {
```
시나리오로 테스트로 애플리케이션이 정말로 올바르게 동작하는지 테스트해 주셨네요. 다만 변화에 좀 취약해 보인다고 느껴졌어요. 각 API들이 변경될 수 있는 여지들이 있는데, 변화에 대응하기 쉬울까? 생각이 들었어요.

메서드 같은 것으로 추상화를 할까 생각이 들기도 했는데, 예를들어서
```java
private void assertTaskNotFound(Long taskId) throws Exception {
    mockMvc.perform(get("/tasks/" + taskId))
            .andExpect(status().isNotFound());
}
```

이렇게 하면 할 일을 불러오는 부분이 변경되면 여기만 변경하고 시나리오에는 영향이 없을수도 있을 것 같아요

EunjinJwa committed:
code review 반영 1

- 각 API의 assertTestCase를 메서드로 분리하고 @Test에서 테스트하고자 하는 목적에 따라 사용하게 함. (개별 테스트, 시나리오 테스트)
- print() 대신 응답 결과를 자동검증 할 수 있도록 수정

EunjinJwa committed:
TEST 코드 모두 작성

EunjinJwa commented:
1. 코드 리뷰 반영
- 각 테스트 단정문을 개별로 작성하여 테스트하고자 하는 곳에서 호출하여 사용함.(개별 테스트, 시나리오 테스트) 
- print() 제거하고 결과 자동 검증 
2. 나머지 Test 코드 작성 

기타 - HelloController MockMvc 테스트 코드 작성하면서 느낀점
- 단순하게 생각했던 테스트인데, HelloController MockMvc 테스트 작성할때 웹 레이어 테스트뿐이라 @WebMvcTest를 사용해보니
TaskService 빈을 못찾는다는 에러가 발생하였습니다. 이 테스트코드를 작성해본 덕분에 어제 단순히 알게되었던 @WebMvcTest 의 스캔 범위와 사용법에 대해 한번 더 찾아보고 이해할 수 있던 것 같습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMockTest.java:
```diff
@@ -0,0 +1,114 @@
+				.andExpect(status().isNoContent());
+	}
+
+}
```
코드 마지막 줄에는 빈 줄이 하나가 추가되어야 합니다. 그렇지 않으면 위와 같이 GitHub에서 표시가 됩니다.
![image](https://github.com/CodeSoom/spring-week3-assignment-1/assets/14071105/8c2bd38e-1d70-4dc8-b4f1-00391d8e00a3)


이 것이 문제가 되는 이유는 POSIX 표준에서 Line을 `개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스`으로 정의하고 있기 때문입니다.

```jsx
// Hello
```

만약 위와 같이 빈 줄이 없는 코드에서 아래에 코드를 추가하게 되면 Git은 다음과 같이 표시합니다.

```jsx
- // Hello
+ // Hello
+ // World
```

저는 단순히 아래 한 줄만 추가했는데 왜 세줄이 변경되었다고 나오는 걸까요? 그 이유는 줄바꿈 문자 `\n`가 들어갔기 때문입니다.

```jsx
- // Hello
+ // Hello\n
+ // World
```

만약 미리 빈 줄을 하나 추가한 상태에서 아래에 새로운 코드를 작성할 경우에는 다음과 같이 보여집니다.

```jsx
+ // World
```

이미 `\n`이 추가되어 있기 때문이죠. 따라서 코드 마지막 줄에 빈 줄 하나를 항상 추가해주는게 좋습니다.

혹은 자동으로 추가할 수 있도록 설정하는 방법도 있습니다. 종립님의 블로그 글을 참고해 주세요

See also

  - [파일-마지막에-자동으로-1개의-개행-문자-추가하기](https://johngrib.github.io/wiki/intellij/#%ED%8C%8C%EC%9D%BC-%EB%A7%88%EC%A7%80%EB%A7%89%EC%97%90-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-1%EA%B0%9C%EC%9D%98-%EA%B0%9C%ED%96%89-%EB%AC%B8%EC%9E%90-%EC%B6%94%EA%B0%80%ED%95%98%EA%B8%B0)

EunjinJwa replied:
아하.. 놓치고 있었네요. 추가했습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,40 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class TaskTest {
```
`Task` 객체에 대해서 아주 꼼꼼하게 테스트를 작성해 주셨네요! 테스트를 우리가 작성한 코드의 예제라고 생각해 봅시다. 테스트가 꼼꼼하게 작성되어 있다면 하나하나의 기능에 대해서는 좋겠지만, 전체적으로 어떻게 사용하는지는 판단하기 어려울 수도 있을 것 같아요. 따라서 이 객체를 어떻게 사용하는지를 간단하게만 작성해 보는 것도 좋을 것 같아요

```java
Task task = new Task();
task.setId(1L);
task.setTitle("책 읽기");

assertThat(task.getId()).isEqualTo(1L);
assertThat(task.getTitle()).isEqualTo("책 읽기");
```

EunjinJwa replied:
그렇네요.. 뭔가 과제가 테스트라서 세세하게 해야할 것 같았어요 ^^;
정작 Task의 사용을 알수있는 테스트는 없었던 것 같네용

hannut91 reviewed (commented):
과제를 제출할 때마다 회고를 하셨군요 ㅎㅎ 제가 아는 어떤 분은 회사에서 일을 할 때  풀 리퀘스트마다 회고를 하시더라구요. 과정을 진행하면서 회고하는 습관도 같이 들이시면 좋겠네요.

hannut91 reviewed (commented):
과제를 제출할 때마다 회고를 하셨군요 ㅎㅎ 제가 아는 어떤 분은 회사에서 일을 할 때  풀 리퀘스트마다 회고를 하시더라구요. 과정을 진행하면서 회고하는 습관도 같이 들이시면 좋겠네요.

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,76 @@
+		assertThat(taskService.getTasks()).hasSize(1);
+	}
+
+	@Test
+	void getTaskWithValidId() {
+		long id = 1L;
+
+		assertThat(taskService.getTask(id)).isNotNull();
+	}
+
+	@Test
+	void getTaskWithInvalidId() {
+		long id = 99L;
+
+		assertThatThrownBy(() -> taskService.getTask(id)).isInstanceOf(TaskNotFoundException.class);
+	}
```
getTask가 주어진 상황(할 일이 존재하는 경우와 존재하지 않는 경우)에 따라 달라지네요. 이런 경우 BDD 테스트 스타일로 테스트를 적으면 주어진 상황을 명시적으로 드러낼 수 있습니다. 예를들어서

```suggestion
@Nested
class Describe_getTask {
    long id = 1L;
    
    @Nested
    class Context_when_task_exsists {
        @BeforeEach
        void setUp() {
            createTaskWithId(id);
        }

	@Test
	void it_returns_task() {
		assertThat(taskService.getTask(id)).isNotNull();
	}
    }

    @Nested
    class Context_when_task_dont_exist {
        @BeforeEach
        void setUp() {
            deleteTaskWithId(1);
        }

	@Test
	void it_returns_task() {
		assertThatThrownBy(() -> taskService.getTask(id)).isInstanceOf(TaskNotFoundException.class);
	}
    }
```

그래서 Describe에는 테스트 대상을
Context에는 주어진 상황을 (그래서 with, when, without 으로만 시작해야 합니다.)

자세한 내용은 종립님 블로그를 한 번 참고해 보세요.

See also
  - https://johngrib.github.io/wiki/junit5-nested/

EunjinJwa replied:
아하.. 생소한 패턴인데, 익숙해지면 상황에 따라 각각 다른 결과를 테스트하기 좋을 것 같습니다.

EunjinJwa committed:
code review 반영

- 마지막 한 줄 공백 라인추가
- TaskTest 사용을 알 수 있도록 테스트코드 변경

EunjinJwa committed:
Describe - Context - It 패턴 적용해봄

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -2,44 +2,100 @@
+		@DisplayName("task가 존재하지 않는다면")
+		class Context_when_tasks_not_exist {
+
+			@BeforeEach
+			void setUpCondition() {
+				deleteTask(SET_ID);
+			}
```
사실 최상위 `@BeforeEach`에서 계속해서 TaskService를 만들고 있어서 사실 안해도 문제는 없긴하지만, 아주 명시적으로 표현하고 싶으면 이렇게 해도 좋은 것 같아요.

hannut91 commented:
테스트를 작성할 때 처음에는 가장 쉽게 성공하는 테스트 케이스와 기능으로 시작하고, 다음에 하나씩 추가해 나가는 방법으로 코드를 작성해 보는 것도 좋아요. 그럴 때 BDD가 좀 잘맞습니다. 예를들어서 할 일 목록을 불러오는 기능을 만든다고 했을 때 할 일 목록을 만드는 기능을 만듭니다. 그 다음에 할 일이 없으면 어떡하지? 이런 것이나, 내림차순으로 요청한다면 어떻게 되지? 이런 것들을 하나씩 추가해 나갈 수 있는데, 자연스럽게 Given을 생각하게돼요

EunjinJwa committed:
spy 객체 테스트 적용해봄
